cpp/src/neighbors/detail/cagra/graph_core.cuh

```diff
@@ -1191,57 +1171,155 @@ void optimize(
-          output_graph_degree,
-          batch_size,
-          i_batch,
-          d_detour_count.data_handle(),
-          d_num_no_detour_edges.data_handle(),
-          dev_stats.data_handle());
+          1024);
+      }
+      const uint32_t batch_size =
+        std::min(static_cast<uint32_t>(graph_size), static_cast<uint32_t>(256 * 1024));
+      const uint32_t num_batch = (graph_size + batch_size - 1) / batch_size;
+      const dim3 threads_prune(32, 1, 1);
+      const dim3 blocks_prune(batch_size, 1, 1);
+
+      RAFT_CUDA_TRY(cudaMemsetAsync(
+        dev_stats.data_handle(), 0, sizeof(uint64_t) * 2, raft::resource::get_cuda_stream(res)));
+
+      for (uint32_t i_batch = 0; i_batch < num_batch; i_batch++) {
+        kern_prune<MAX_DEGREE, IdxT>
+          <<<blocks_prune, threads_prune, 0, raft::resource::get_cuda_stream(res)>>>(
+            d_input_graph.data_handle(),
+            graph_size,
+            input_graph_degree,
+            output_graph_degree,
+            batch_size,
+            i_batch,
+            d_detour_count.data_handle(),
+            d_num_no_detour_edges.data_handle(),
+            dev_stats.data_handle());
+        raft::resource::sync_stream(res);
+        RAFT_LOG_DEBUG(
+          "# Pruning kNN Graph on GPUs (%.1lf %%)\r",
+          (double)std::min<IdxT>((i_batch + 1) * batch_size, graph_size) / graph_size * 100);
+      }
       raft::resource::sync_stream(res);
-      RAFT_LOG_DEBUG(
-        "# Pruning kNN Graph on GPUs (%.1lf %%)\r",
-        (double)std::min<IdxT>((i_batch + 1) * batch_size, graph_size) / graph_size * 100);
-    }
-    raft::resource::sync_stream(res);
-    RAFT_LOG_DEBUG("\n");
+      RAFT_LOG_DEBUG("\n");
 
-    host_matrix_view_from_device<uint8_t, int64_t> detour_count(res, d_detour_count.view());
+      raft::copy(detour_count.data_handle(),
+                 d_detour_count.data_handle(),
+                 detour_count.size(),
+                 raft::resource::get_cuda_stream(res));
 
-    raft::copy(
-      host_stats.data_handle(), dev_stats.data_handle(), 2, raft::resource::get_cuda_stream(res));
-    const auto num_keep = host_stats.data_handle()[0];
-    const auto num_full = host_stats.data_handle()[1];
+      raft::copy(
+        host_stats.data_handle(), dev_stats.data_handle(), 2, raft::resource::get_cuda_stream(res));
+      num_keep = host_stats.data_handle()[0];
+      num_full = host_stats.data_handle()[1];
+    } else {
+      // Count 2-hop detours on CPU
+      raft::common::nvtx::range<cuvs::common::nvtx::domain::cuvs> block_scope(
+        "cagra::graph::optimize/prune/2-hop-counting-by-CPU");
+
+#pragma omp parallel for reduction(+ : num_keep, num_full)
+      for (IdxT iA = 0; iA < graph_size; iA++) {
+        uint32_t num_edges_no_detour = 0;
+        // Create a list of nodes, iB_candidates, that can be reached in 2-hops from node A.
+        auto iB_candidates = raft::make_host_vector<IdxT, int64_t>((input_graph_degree - 1) *
```

Comment:
Could you describe the layout of this vector? If I understand correctly, the elements are sorted so that index range `0..k^2-1` store the elements that can be accessed by considering only the first `k` neighbors of the nodes in the graph.

Reply:
`iB_candaites` is a temporary array that holds the node IDs reached from node-A by 2-hop routes. If degree of the knn graph is D, the number of 2-hop routes is `D^2`, but the size of `iB_candidates` is `(D-1)^2`. Why? Because if the rank of an edge from node-A is R `[0:D-1]`, then the only 2-hop routes that qualify as 2-hop detours for this edge are those that use two edges, both of which have ranks less than R. Therefore, the maximum number of possible 2-hop detours is `(D-1)^2`, which is why the size of `iB_candidates` is `(D-1)^2`.

Reply:
I think we can skip this note, the way the data is assigned to the array describes the layout.